src/zh/happymh/src/eu/kanade/tachiyomi/extension/zh/happymh/Happymh.kt

```diff
@@ -194,25 +202,49 @@ class Happymh : HttpSource(), ConfigurableSource {
             .map {
                 // remove order mark
                 it.onEach { chapter ->
-                    chapter.chapter_number = 0f
+                    chapter.chapter_number = -1f
                 }
             }
     }
 
     override fun chapterListParse(response: Response) = throw UnsupportedOperationException()
 
     override fun getChapterUrl(chapter: SChapter): String {
-        return baseUrl + chapter.url
+        return baseUrl + (chapterUrlToCode.get(chapter.url)?.let { "/mangaread/$it" } ?: chapter.url)
     }
 
     // Pages
 
+    private fun fetchChapterCode(chapter: SChapter): String? {
+        val expectPage = chapter.url.substringAfterLast("#").toIntOrNull() ?: 1
+        val pathSegments = chapter.url.substringBeforeLast("#").split("/")
+        val comicId = pathSegments[1]
+        val chapterId = pathSegments[3].toLong()
+        var code = fetchChapterByPage(comicId, expectPage).items.find { it.id == chapterId }?.codes
+        if (code == null) {
+            // Do full search for find target code
+            var page = 1
+            var end = false
+            while (!end && code == null) {
+                val resp = fetchChapterByPage(comicId, page)
+                code = resp.items.find { it.id == chapterId }?.codes
+                end = resp.isPageEnd()
+                page += 1
+            }
```

Comment:
The `takeWhile `method discards items that do not meet the specified condition. In this case, I need a `takeUntil `method, but it seems that Kotlin does not provide one. If I were to implement it specifically, I think it would be better not to use the `Stream `form at all.